### PR TITLE
feat: Style-Controlled Canvas Dimensions

### LINKED ISSUE #361: Canvas size shouldn't be hardcoded
Canvas dimensions are currently set to 800 x 700 in `Canvas.tsx`. This is noticeable when downloading the SVG from the relevant button, since the downloaded SVG crops out everything outside of those dimensions.

### PR DESCRIPTION
**Pull request by joshpoll**
# Description

Closes #361.

Currently the canvas size is fixed to 800x700 pixels. This PR requires that users specify the dimensions of the canvas in Style. Specifically, the values `canvas.width` and `canvas.height` in Style are used to control the canvas dimensions. The canvas dimensions must be numeric literals. For example, one may write

```
canvas {
  width = 800
  height = 700
}
```

This is a breaking change, requiring an explicit definition of canvas dimensions in Style.

This PR also introduces one more breaking change. Currently the `maxSize` constraint relies on the size of the canvas to determine the maximum size of an object. This PR changes the API of `maxSize` to `maxSize(variable, bound)`. The previous behavior can be obtained by writing `maxSize(variable, min(canvas.width, canvas.height))`. I made this change, because it is difficult to thread canvas dimensions into constraints, and because `maxSize` is likely to change in the future anyway: #424. However, I can look into this issue more closely and see if I can incorporate canvas dimensions into constraints.

# Implementation strategy and design decisions

We considered several options for incorporating canvas size including adding frontmatter, a new file type, or a new Canvas pseudo-shape. Ultimately we decided to go with this simple global variable approach, because we just want to :shipit:, and these other concerns will likely resurface at a later date when we can address them with their own PR.

Canvas size is set in the Style program and initialized, from the translation, in the `State` output by the Style compiler. The `canvas` parameter in `State` is passed around to anything that might need it, such as various shape/property samplers/resamplers.

# Checklist

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass locally using `yarn test`
- [x] I ran `yarn docs` and there were no errors when generating the HTML site
- [x] My code follows the style guidelines of this project (e.g.: no ESLint warnings)

# TODOs
- [x] Remove remaining hard-coded occurrences of canvas dimension.
- [x] Add error-handling when canvas dimensions are not set properly in Style.
- [x] Add tests for error-handling.
- [x] Delete experimental `.sty`.

# Open questions

- Should there be a default canvas size? We have decided "no," since there is no obvious default to set it to. The current baked-in value of 800x700 is essentially arbitrary.
- Should `maxSize` have a new parameter? `maxSize` has informally been deprecated, so its interface is unstable.

### CODE REVIEW

**Comment by hypotext** [comment]
Did we perhaps want a default canvas size? I can also see the case for always having the Style writer explicitly specify it.

**Comment by joshpoll** [comment]
I lean towards explicit canvas dimensions, but let's discuss in #lang-design

**Comment by joshpoll** [comment]
After discussing in #lang-design, we have decided to keep canvas size explicit for now.

**Comment by hypotext** [comment]
Requesting a quick once-over from @maxkrieger for inspector-related changes

**Review by hypotext — APPROVED** [review]
Thanks for doing this, it's a much needed change! Looks good overall. I checked the code related to the Style compiler, and will leave inspector/React-related review to @wodeni and @maxkrieger to approve.

My main question is: Did you confirm that all existing examples work? I remember that @wodeni's automator auto-generates images of all examples in `examples/registry.json` (though I don't remember where the example images are saved). Maybe you can confirm those look fine. 

Another question: It would be good to clarify in a spec somewhere the intended behavior of `canvas` WRT the rest of the Style file. Does `canvas` need to be at the start of the Style file, or can it be anywhere? Can `canvas` be overridden again in the Style file after it's defined, just like any other namespace?

After this PR is merged, I would request that you do the following:

- Edit wiki/tutorial examples to use/discuss `canvas` and `maxSize` edits for the Style writer where relevant (while tedious, keeping the "hot" docs up-to-date is an important part of extending an evolving PL)
  - https://github.com/penrose/penrose/wiki/Getting-started
  - https://github.com/penrose/penrose/wiki/Style-language-spec
  - https://penrose.gitbook.io/penrose/

Also, we should probably do a version bump after this PR is merged (I can do it, cc @wodeni)

Note that I edited your PR description to add some more about the implementation strategy, which is going to be useful for future people who want to know what the role of `canvas` in the system. I'd encourage documentation at this level.

**Comment by hypotext** [comment]
It would also be nice to have a simple example with very big vs. very small canvas size, and to show how the shapes are laid out differently (with a line drawn around the canvas border). This would be a nice demo to add to our Examples wiki page.

File: packages/core/src/utils/Error.ts
Comment (wodeni) [reply packages/core/src/utils/Error.ts:341]: Ahh beautiful!

File: packages/core/src/compiler/Style.ts
Comment (joshpoll) [reply packages/core/src/compiler/Style.ts:3163]: Yeah I normally try to avoid casts. The way to remove the casts here would be to add a `checkCanvas` call at the top and hope TypeScript can infer correctly based on that. However, `checkCanvas` is already run before any call to `getCanvas`.

File: packages/core/src/compiler/Style.ts
Comment (joshpoll) [reply packages/core/src/compiler/Style.ts:3083]: oops I think this comment is stale. I'll remove it

File: packages/core/src/compiler/Style.ts
Comment (joshpoll) [reply packages/core/src/compiler/Style.ts:3163]: Ok so TS is unable to infer the correct types even when I add a `checkCanvas` call to `getCanvas`, so I think the casts are unavoidable.

**Comment by joshpoll** [comment]
Thanks for the feedback! I will add some examples with differing canvas sizes, and update the docs/wiki. To clarify: did you mean that the canvas spec should also go in the docs/wiki or somewhere in a doc comment in the code?

**Review by maxkrieger — APPROVED** [review]
LGTM! great PR! appreciate your integration with all our prior dataflows like canvasSize

File: packages/core/src/utils/Error.ts
Comment (joshpoll) [reply packages/core/src/utils/Error.ts:341]: thanks!

File: packages/core/src/compiler/Style.ts
Comment (joshpoll) [reply packages/core/src/compiler/Style.ts]: done

**Comment by joshpoll** [comment]
update:
- I have documented both the canvas and maxSize changes on the wiki
- I have added large and small canvas examples
- I have confirmed that existing examples still work as intended

**Comment by joshpoll** [comment]
Post-merge I will update the tutorial as well

File: packages/core/src/compiler/Style.ts
Comment (hypotext) [reply packages/core/src/compiler/Style.ts:3128]: I mean that in the height case, there is extra code that does something different (L3144-3155) from the width case ) (L3114-3118). I'd pick one approach and stick with it.

**Comment by hypotext** [comment]
> To clarify: did you mean that the canvas spec should also go in the docs/wiki or somewhere in a doc comment in the code?

Docs/wiki is fine.

> I have added large and small canvas examples

Can you add working paths/images for the examples here? https://github.com/penrose/penrose/wiki/Example-diagrams

Since Max approved it, after we address https://github.com/penrose/penrose/pull/589#discussion_r652951822 then I can merge this PR!

File: packages/core/src/compiler/Style.ts
Comment (joshpoll) [reply packages/core/src/compiler/Style.ts:3128]: Ah I see now! The longer one is correct (and is what the stale comment was referring to), but I forgot to change both. I have updated the `width` path to match. Thanks for pointing that out!

**Review by wodeni — APPROVED** [review]
Looking great! The changes to `ShapeDef` look good to me. 

On `getCanvas`: it's probably better to (a) check the canvas block is well-formed before the full `Translation` is generated and (b) store the `Canvas` value in a specifcal place (e.g. a specific attribute of `State`?), instead of `Translation`. Not sure if (a) and (b) are both required to resolve the `getCanvas` issue, but I think it's not that urgent. We can wait until there are more of these preludes in the system at some point.

**Comment by joshpoll** [comment]
The first point is a good one actually. I'm not sure how early the check can be run. It has to be run after e.g. `override` is resolved I think, but I don't know where that happens. The second issue doesn't apply I think, because I added `Canvas` to `State`. Before that (i.e. in `getCanvas`) the canvas properties are looked up in `Translation` dynamically and not stored in any special way.

I agree that if there are more special properties like canvas, we'd ideally have a more robust solution.

**Pull request merged**
